src/hotspot/share/cds/heapShared.cpp

```diff
@@ -781,6 +781,137 @@ void HeapShared::copy_objects() {
   }
 }
 
+// Mark the classes of all objects that are scanned HeapShared::find_all_aot_initialized_classes()
+class HeapShared::AOTInitializedClassScanner : public BasicOopIterateClosure {
+  bool _made_progress;
+
+  template <class T> void check(T *p) {
+    oop obj = HeapAccess<>::oop_load(p);
+    if (!java_lang_Class::is_instance(obj)) {
+      // Don't scan the mirrors, as we may see an orig_mirror while scanning
+      // the object graph, .... TODO more info
+      _made_progress |= HeapShared::scan_for_aot_initialized_classes(obj);
+    }
+  }
+
+public:
+  AOTInitializedClassScanner() : _made_progress(false) {}
+  void do_oop(narrowOop *p) { check(p); }
+  void do_oop(      oop *p) { check(p); }
+  bool made_progress() { return _made_progress; }
+};
+
+static bool mark_for_aot_initialization(InstanceKlass* buffered_ik) {
+  if (buffered_ik->name()->equals("java/lang/String")) {
+    return false;
+  }
+
+  assert(ArchiveBuilder::current()->is_in_buffer_space(buffered_ik), "sanity");
+
+  bool made_progress = false;
+  if (buffered_ik->is_initialized() && !buffered_ik->has_aot_initialized_mirror()) {
+    if (log_is_enabled(Info, cds, init)) {
+      ResourceMark rm;
+      log_info(cds, init)("Mark class for aot-init: %s", buffered_ik->external_name());
+    }
+
+    buffered_ik->set_has_aot_initialized_mirror();
+    made_progress = true;
+
+    InstanceKlass* super = buffered_ik->java_super();
+    if (super != nullptr) {
+      mark_for_aot_initialization(super);
+    }
+
+    Array<InstanceKlass*>* interfaces = buffered_ik->local_interfaces();
+    for (int i = 0; i < interfaces->length(); i++) {
+      mark_for_aot_initialization(interfaces->at(i));
+    }
+  }
+
+  return made_progress;
+}
+
+void HeapShared::find_all_aot_initialized_classes() {
+  if (!CDSConfig::is_dumping_aot_linked_classes()) {
+    return;
+  }
+
+  init_seen_objects_table();
+
+  GrowableArray<Klass*>* klasses = ArchiveBuilder::current()->klasses();
+  assert(klasses != nullptr, "sanity");
+
+  // First scan all resolved constant pools references.
+  for (int i = 0; i < klasses->length(); i++) {
+    Klass* src_k = klasses->at(i);
+    if (src_k->is_instance_klass()) {
+      InstanceKlass* src_ik = InstanceKlass::cast(src_k);
+      InstanceKlass* buffered_ik = ArchiveBuilder::current()->get_buffered_addr(src_ik);
+      objArrayOop rr = buffered_ik->constants()->prepare_resolved_references_for_archiving();
+      if (rr != nullptr) {
+        for (int i = 0; i < rr->length(); i++) {
+          scan_for_aot_initialized_classes(rr->obj_at(i));
+        }
+      }
+
+      if (AOTClassInitializer::can_archive_initialized_mirror(src_ik)) {
+        mark_for_aot_initialization(buffered_ik);
+      }
+    }
+  }
+
+  // These objects also belong to the special subgraph
```

Comment:
Can you add a "why" to this comment?  I think it's because we preserve them specially in universe.cpp::BuiltinException but tying these two things together with a comment would help.

Assuming the above is correct, it would be useful to have a way to automatically discover new `BuiltinException`s so they also get AOT initialized without having to remember to update this code.

Have you considered using an implicit list to link them together so there would be a single BuiltInException `head` that could be used to walk the list and find all the relevant `BuiltinException` instances?